README.md
![A FordLabs Product](images/FluxHero.png)

# Flux

Flux is a flexible timer for establishing your working rhythm.

## Background

Psychology defines _flow_ as the mental state in which you are fully immersed in an activity. This is sometimes referred to as being "in the zone." Managing your flow is an important part of optimizing your productivity.

Over the years, three different work rhythms have emerged in pursuit of managing flow: Pomodoro, Desktime, and Ultradian.

1. Pomodoro utilizes 25 minutes of work with a 5 minute break.
2. Desktime utilizes 52 minutes of work with a 17 minutes break.
3. Ultradian utilizes 90 minutes of work with an extended break.

This extension provides a timer that can be configured to operate at any of these rhythms.

> Tip: This [blog post](https://medium.com/@timmetz/pomodoro-technique-and-other-work-rhythms-which-one-suits-you-34c2d05fe46e) provides a great overview of these work rhythms.

## Getting Started

To get started, activate this extension by running the VS Code command `Flux`.

> Hint: To run a command in VS Code hit **cmd/ctrl + shift + p**

This will create an item on the left side of your status bar. Click the item when you are ready to begin the timer.

## Features

### Pausing

Clicking the timer while it is running will pause/unpause it.

### Resetting

To reset the timer, run the VS Code command `Reset Flux`.

### Configure your working rhythm

Flux provides a rhythm setting that will allow you to select between the Pomodoro, Desktime, and Ultradian rhythms.

> Hint: Flux will detect your configured rhythm upon extension activation, as well as when you run the reset command.

## Requirements

This extension is designed to work right out of the box.

## Release Notes

Flux is in public beta. We are hard at work creating the best experience possible for this extension. If you have any feedback, feel free to submit an issue to this repository.

### 0.0.1

Initial release of Flux

## License

Flux is licensed under the Apache License 2.0. See [LICENSE](LICENSE) for details.

## Attribution

Hourglass icon made by [Smashicons](https://smashicons.com/) from [flaticon](https://www.flaticon.com/)


CHANGELOG.md
# Change Log

All notable changes to the "flux" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

src/test/timer/timer.test.ts
import Timer from '../../timer/timer';

describe('Timer', () => {
  let timer: Timer;

  beforeEach(() => {
    timer = new Timer();
  });

  it('should have a default duration of 0', () => {
    expect(timer.duration.asMilliseconds()).toEqual(0);
  });

  describe('setTimeRemaining', () => {
    it('should update the duration', () => {
      timer.setTimeRemaining(3);
      expect(timer.duration.asMilliseconds()).toEqual(180000);
    });
  });

  describe('getTimeRemaining', () => {
    it('should get the total time as milliseconds', () => {
      timer.setTimeRemaining(7);
      expect(timer.getTimeRemaining().total).toEqual(420000);
    });

    it('should get the minutes as padded string', () => {
      timer.setTimeRemaining(7);
      expect(timer.getTimeRemaining().minutes).toEqual('07');
    });

    it('should get the seconds as padded string', () => {
      timer.setTimeRemaining(7);
      expect(timer.getTimeRemaining().seconds).toEqual('00');
    });

    it('should subtract a second', () => {
      timer.setTimeRemaining(7);
      timer.getTimeRemaining();
      const updatedTime = timer.getTimeRemaining();
      expect(updatedTime.minutes).toEqual('06');
      expect(updatedTime.seconds).toEqual('59');
    });
  });
});


src/extension.ts
import { ExtensionContext, commands } from 'vscode';
import TimerService from './timer/timerService';
import { Widget, StatusBarCommandId } from './widget';

let timerService = new TimerService();

export function activate(context: ExtensionContext) {
  console.log('Congratulations, your extension "flux" is now active!');

  let command = commands.registerCommand(StatusBarCommandId, () => {
    if (timerService.hasStarted) {
      timerService.isPaused = !timerService.isPaused;
    } else {
      timerService.beginTimer();
    }
  });

  context.subscriptions.push(command);

  commands.registerCommand('extension.flux', () => {
    context.subscriptions.push(Widget.Instance.statusBarItem);
  });

  commands.registerCommand('extension.resetFlux', () => {
    timerService.resetTimer();
    Widget.Instance.statusBarItem.text = `$(watch) Click to Start Flux Timer`;
  });
}

export function deactivate() {}


src/timer/timerService.ts
import { window, workspace } from 'vscode';
import { Widget } from '../widget';
import Timer from './timer';

let interval: NodeJS.Timeout;

export default class TimerService {
  timer: Timer;
  hasStarted: boolean = false;
  isPaused: boolean = false;

  constructor() {
    this.timer = new Timer();
  }

  beginTimer() {
    this.timer.setTimeRemaining(this.getConfiguredTime());
    this.updateTimerValue(this.timer.getTimeRemaining());

    interval = setInterval(() => {
      if (!this.isPaused) {
        const timeRemaining = this.timer.getTimeRemaining();
        this.updateTimerValue(timeRemaining);
      }
    }, 1000);

    this.hasStarted = true;
    window.showInformationMessage('Your Flux Timer Has Started');
  }

  resetTimer() {
    clearInterval(interval);
    this.hasStarted = false;
  }

  private endTimer() {
    this.resetTimer();
    window.showInformationMessage('Time to Take a Break');
    Widget.Instance.statusBarItem.text = `$(watch) 00:00 - Click to Restart Flux Timer`;
  }

  private updateTimerValue(timeRemaining: any) {
    if (timeRemaining.total < 0) {
      this.endTimer();
    } else {
      Widget.Instance.statusBarItem.text = `$(watch) ${timeRemaining.minutes}:${timeRemaining.seconds}`;
    }
  }

  private getConfiguredTime(): number {
    const rhythm = workspace.getConfiguration('flux').get('rhythm');
    switch (rhythm) {
      case 'Pomodoro':
        return 25;
      case 'Desktime':
        return 52;
      case 'Ultradian':
        return 90;
      default:
        return 25;
    }
  }
}
